validate block create cli output

oshankkumar opened the pull request:
- defined struct for block create response
- validate the cli output
Signed-off-by: Oshank Kumar <okumar@redhat.com>

Commit:
validate block create cli output

- defined struct for block create response
- validate the cli output
- accept block size in no. of bytes or [KMGT]iB or [KMGT]B format

Signed-off-by: Oshank Kumar <okumar@redhat.com>

Madhu-1 commented on glusterblockrestd/blockvolmanager/volume_manager.go:
```diff
@@ -0,0 +1,144 @@
+			return nil, err
+		}
+		blockSizeInBytes = uint64(sz)
+	case float64:
```
is this required? I don't think will support float size

oshankkumar replied:
go unmarhsal any number(int/float) in float64 by default, since i used size as interface{} type, so in type switch i used float64

Madhu-1 commented on glusterblockrestd/blockvolmanager/volume_manager.go:
```diff
@@ -0,0 +1,144 @@
+
+	switch blockSize := req.Size.(type) {
+	case string:
+		sz, err := size.Parse(blockSize)
```
this is an input request validation it better to do it in HTTP request handler.

Madhu-1 commented on pkg/api/api.go:
```diff
@@ -10,13 +10,34 @@ type ErrorResp struct {
-	RingBufferSizeInMB uint64   `json:"ring-buffer-size-mb"`
-	Hosts              []string `json:"hosts"`
+	// Name represents block Volume name
+	Name               string      `json:"name,omitempty"`
```
the name should be required param

Madhu-1 commented on pkg/api/api.go:
```diff
@@ -10,13 +10,34 @@ type ErrorResp struct {
-	Hosts              []string `json:"hosts"`
+	// Name represents block Volume name
+	Name               string      `json:"name,omitempty"`
+	HostingVolume      string      `json:"hosting_volume,omitempty"`
```
This is not decided yet, whether the glusterd blockd rest server will take care of handing block hosting volume creation and retrieval.

oshankkumar replied:
gluster-block cli requires hosting volume name also for block vol creation

Madhu-1 replied:
then its should not be omitempty

Madhu-1 commented on pkg/api/api.go:
```diff
@@ -10,13 +10,34 @@ type ErrorResp struct {
+	HaCount            int         `json:"ha,omitempty"`
+	AuthEnabled        bool        `json:"auth-enabled"`
+	FullPrealloc       bool        `json:"full-prealloc"`
+	Size               interface{} `json:"size,omitempty"`
```
make size as uint64 and it should accept in only bytes. I don't think we need to support all other units

Madhu-1 commented on pkg/api/api.go:
```diff
@@ -10,13 +10,34 @@ type ErrorResp struct {
+	Size               interface{} `json:"size,omitempty"`
+	Storage            string      `json:"storage,omitempty"`
+	RingBufferSizeInMB uint64      `json:"ring-buffer-size-mb,omitempty"`
+	Hosts              []string    `json:"hosts,omitempty"`
```
Hosts is not a required param for block volume creation?

oshankkumar replied:
Hosts are used in gluster-block cli

Madhu-1 replied:
why its omitempty?

Madhu-1 commented on pkg/size/size.go:
```diff
@@ -0,0 +1,184 @@
+package size
```
This file is not needed as we need to support only volume size in bytes.

Madhu-1 commented on pkg/size/size_test.go:
```diff
@@ -0,0 +1,160 @@
+package size
```
same here as well

Madhu-1 commented on pkg/api/api.go:
```diff
@@ -10,13 +10,34 @@ type ErrorResp struct {
+// BlockVolumeCreateError respresents error in creating a
+// gluster block volume
+// Note: same as gluster-block cli output
+type BlockVolumeCreateError struct {
```
this can be the generic error structure, I think this will be useful for other operations also

Commit:
refactor handler code

Signed-off-by: Oshank Kumar <okumar@redhat.com>

oshankkumar commented:
Refactored some of code in handler package and also renamed the file common.go to routes.go, since it contains all routing related info

Commit:
added interface for all block vol management operation

- added hosting vol name and vol name in path parameter,
  these params should be mandatory

Signed-off-by: Oshank Kumar <okumar@redhat.com>

Commit:
added debug log to print gluster-block cli command being executed

Signed-off-by: Oshank Kumar <okumar@redhat.com>

Madhu-1 commented on glusterblockrestd/handlers/routes.go:
```diff
@@ -0,0 +1,38 @@
+	"net/http"
+
+	"github.com/gluster/gluster-block-restapi/glusterblockrestd/blockvolmanager"
+	"github.com/gorilla/mux"
```
new line here

oshankkumar replied:
done

Madhu-1 commented on pkg/utils/utils.go:
```diff
@@ -49,21 +52,30 @@ func ExecuteCommandOutput(cmdName string, arg ...string) ([]byte, error) {
 // ExecuteCommandRun runs the command and adds additional
 // error information
-func ExecuteCommandRun(cmdName string, arg ...string) error {
+func ExecuteCommandRun(cmdName string, arg ...string) (err error) {
```
why this is names return?

oshankkumar replied:
I will remove that,
